### PR TITLE
Adds vault role to JWT claims if specified in jobspec

### PR DESCRIPTION
**Pull request by mikenomitch**
Adds vault_role as an attribute in the JWT if, and only if, the role is specified in the jobspec. This would allow Vault users to interpolate the Vault role in their templated policies. This could be generally useful, but was primarily added to unblock a specific customer. They have verified that this fix (including the constraint below) would unblock JWT-based Vault auth for their setup.

Ideally, we could add the role regardless of where it is specified, but since the server is minting the JWTs and the client has the default fallback auth_role values, it is difficult to insert the role into the JWT claims if it is using the fallback. There is also a potential syncing issue if you update config and then restart Nomad. To get around these issues, Vault Namespaces have the same constraint (only adding to claims if in the jobspec), so we're adding the same constraint to the role JWT claim.

### CODE REVIEW

File: .changelog/19535.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+identity: Added vault role to jwt claims if specified in jobspec
```
Comment (lgfa29) [inline .changelog/19535.txt:2]:
```suggestion
identity: Added `vault_role` to JWT workload identity claims if specified in jobspec
```

File: nomad/structs/workload_id.go [hunk 2]
```diff
@@ -302,6 +302,7 @@ type WIHandle struct {
 	IdentityName string
 	// WorkloadIdentifier is either a ServiceName or a TaskName
 	WorkloadIdentifier string
+	VaultRole          string
```
Comment (lgfa29) [inline nomad/structs/workload_id.go:305]:
```suggestion
```

I don't think we need this anymore?

File: website/content/docs/concepts/workload-identity.mdx [hunk 3]
```diff
@@ -38,6 +38,23 @@ instead of the task name.
 }
 ```
 
+Workload identities for Vault services have an additional claim for
+[`vault.role`][jobspec_vault_role] if role is specified in the job.
```
Comment (lgfa29) [inline website/content/docs/concepts/workload-identity.mdx:42]:
```suggestion
Workload identities for tasks that use Vault have an additional claim for
[`vault.role`][jobspec_vault_role] if a role is specified in the job.
```

**Review by lgfa29 — APPROVED** [review]
LGTM! Mostly some minor suggestions, and I think we need to remove that `VaultRole` from `WIHandle` since we're not using it anymore?

File: nomad/structs/workload_id.go (hunk 2 above)
Comment (mikenomitch) [reply nomad/structs/workload_id.go:305]: Oh damn, sorry I missed this 🤦- thought this was the Claims struct when I was self-reviewing

**Comment by BGMUSTC** [comment]
When would this feature released? Do you have any plan?

**Comment by pkazmierczak** [comment]
@BGMUSTC The feature will be released in the upcoming 1.7.3 release.

**Pull request merged**